### ISSUE TITLE
For Consideration: Add support to replace "libde_rest_plugin.so" with custom binary.

### ISSUE THREAD

**Issue by ericvandamme** [marthoc/docker-deconz#114]
It would be a nice to have feature to be able to replace the "libde_rest_plugin.so" plugin with one custom built between releases. I am working with some other people deploying their Home Assistant, but occasionally support for new sensors / lights need to be added.

**Comment by marthoc** [comment]
I think you should be able to do this already, just mount your custom compiled libde_rest_plugin.so into the container:
-v /path/to/your/libde_rest_plugin.so:/usr/share/deCONZ/plugins/libde_rest_plugin.so

**Comment by Krocko** [comment]
If i specify the path to the rest plugin, the custom path for the persist storage (-v /opt/deconz:/root/.local/share/dresden-elektronik/deCONZ \) don't work anymore. Or is ignored.

**Comment by Krocko** [comment]
@marthoc any help on this please?

**Comment by Krocko** [comment]
Anyone got this working?
Would be really good to have this feature.

**Comment by SAOPP** [comment]
I'm also started to compile latest commits before next release, 'coz I want to pair my brand new opple switches... and also confused faced with some procedure to replace plugin library. Not sure where need to move.

**Comment by MarcProe** [comment]
After compiling, I tried to:
```
docker cp ../libde_rest_plugin.so docker:/tmp
sudo docker exec -it deconz bash
cp /tmp/libde_rest_plugin.so /usr/share/deCONZ/plugins
```

After restarting the container, the plugin is simply ignored: 
```
20:22:30:985 HTTP Server listen on address 0.0.0.0, port: 80, root: /usr/share/deCONZ/webapp/
20:22:31:008 CTRL. 3.16.220:22:31:340 dev /dev/ttyAMA0
20:22:31:340 ZCLDB init file /root/.local/share/dresden-elektronik/deCONZ/zcldb.txt
20:22:31:436 found node plugin: libde_signal_plugin.so - Signal Monitor Plugin
20:22:31:453 found node plugin: libstd_otau_plugin.so - STD OTAU Plugin
20:22:31:727 dev /dev/ttyAMA0
```

**Comment by lukicsl** [comment]
This is strange I forked the git project which creates the docker container (marthoc/deconz) and adapted the Dockerfile

```
.......
COPY libde_rest_plugin.so /usr/share/deCONZ/plugins

VOLUME [ "/root/.local/share/dresden-elektronik/deCONZ" ]

EXPOSE ${DECONZ_WEB_PORT} ${DECONZ_WS_PORT} ${DECONZ_VNC_PORT}

ENTRYPOINT [ "/tini", "--", "/start.sh" ]

```
Which gives 0 results the login to RESt and Phoscon does not work

**Comment by lukicsl** [comment]
I do not understand why this does not work. I created a Dockerfile:
```
FROM marthoc/deconz
COPY libde_rest_plugin.so /usr/share/deCONZ/plugins

```
I verified that the file is copied to the container. Still does nor work. I am not able to login to Phoscon with the credential I used before.

**Comment by SAOPP** [comment]
Well guys, I previously reported that I did not succeed with replacing the plugin with a newly compiled one. Damn it, I'm tired of waiting, I have 6 switches already :))))

**Comment by Rick-Jongbloed** [comment]
For the people using docker on a Raspberry Pi, i managed to compile the plugin from master on the docker image:

Start a shell in the docker image, then run
```
cd /opt
apt update
apt install qt5-default libqt5sql5 libqt5websockets5-dev libqt5serialport5-dev libsqlite3-dev build-essential git
git clone https://github.com/dresden-elektronik/deconz-rest-plugin.git
wget http://deconz.dresden-elektronik.de/ubuntu/beta/deconz-dev-2.05.75.deb
dpkg -x deconz-dev-2.05.75.deb ./deconz-dev
cp ./deconz-dev/usr/include/* /usr/include/ -R
qmake && make -j2
```

now the plugin ../libde_rest_plugin.so is created. Extract the plugin from the docker image and mount it to the image using the -v parameter
-v /path/to/your/libde_rest_plugin.so:/usr/share/deCONZ/plugins/libde_rest_plugin.so`
(I got this command from this thread, didn't test it bc i'm using Kubernetes)

After restarting the container the image should be refreshed and the new plugin is working, without any authentication issues to Phoscon, VNC or the API.

**Comment by lukicsl** [comment]
> wget http://deconz.dresden-elektronik.de/ubuntu/beta/deconz-dev-2.05.75.deb
> dpkg -x deconz-dev-2.05.75.deb ./deconz-dev
> cp ./deconz-dev/usr/include/* /usr/include/ -R
> qmake && make -j2

Compilation fails:
```
/usr/bin/ld: cannot find -ldeCONZ
collect2: error: ld returned 1 exit status
Makefile.Release:321: recipe for target '../libde_rest_plugin.so' failed
make[1]: *** [../libde_rest_plugin.so] Error 1
make[1]: Leaving directory '/home/lukics/git.com/deconz-rest-plugin'
Makefile:38: recipe for target 'release' failed
make: *** [release] Error 2
```
Instead you can do (take care about the architecture, which I did not do in the past):
```
wget -O - http://phoscon.de/apt/deconz.pub.key | \
           sudo apt-key add -

sudo sh -c "echo 'deb [arch=armhf] http://phoscon.de/apt/deconz \
            $(lsb_release -cs)-beta main' > \
            /etc/apt/sources.list.d/deconz.list"
sudo apt update
sudo apt install deconz-dev

```
I had an issue before because I did not see in the command above that [arch] has to be set and  I got the wrong package. Now the library is picked up and loaded

**Comment by lukicsl** [comment]
The complete steps I did on my Pi (armhf):
```
cd
mkdir git.com
cd git.com
git clone https://github.com/dresden-elektronik/deconz-rest-plugin.git
cd deconz-rest-plugin
wget -O - http://phoscon.de/apt/deconz.pub.key | \
           sudo apt-key add -

sudo sh -c "echo 'deb [arch=armhf] http://phoscon.de/apt/deconz \
            $(lsb_release -cs)-beta main' > \
            /etc/apt/sources.list.d/deconz.list"
sudo apt update
sudo apt install deconz-dev
qmake && make -j2
```
I am using docker-compose and defined for all my containers a mountpoint at /datavol.
For Phoscon it is:
```
/datavol/deconz/conf 
/datavol/deconz/lib
```
I copied the compiled file to the mountpoint
`cp ../libde_rest_plugin.so /datavol/deconz/lib`

the docker-compose.yml is:
```
version: "2"
services:
  deconz:
    image: marthoc/deconz:latest
    container_name: deconz
    network_mode: host
    restart: always
    volumes:
      - /datavol/deconz/conf:/root/.local/share/dresden-elektronik/deCONZ
      - /datavol/deconz/lib/libde_rest_plugin.so:/usr/share/deCONZ/plugins/libde_rest_plugin.so
    devices:
      - /dev/ttyAMA0
    environment:
      - DECONZ_WEB_PORT=80
      - DECONZ_VNC_MODE=1
      - DECONZ_VNC_PASSWORD=.guesswhat.
      - DECONZ_WS_PORT=443
      - DEBUG_INFO=1
      - DEBUG_APS=0
      - DEBUG_ZCL=0
      - DEBUG_ZDP=0
      - DEBUG_OTAU=0

```

**Comment by Guus-H** [comment]
> The complete steps I did on my Pi (armhf):
> 

So for the newbs (like me) on the block, Raspberry Pi 4, that should normally run armhf? I have googled it, but haven't found a straightforward answer..

**Comment by Rick-Jongbloed** [comment]
> Raspberry Pi 4, that should normally run armhf?

Yes

**Comment by Rick-Jongbloed** [comment]
While my solution does work on the docker image maintained by @marthoc, I tried to replicate the correct way to install the deconz headers using the deCONZ apt repo on the Docker image.

I had to install git, wget build-essential and lsb-release first:
`apt update && apt install wget git lsb-release build-essential`
then the instruction posted by @lukicsl can be used.

**Comment by smesguich** [comment]
Hello !
I'm facing the same issue.
I'm running latest deCONZ docker on my Synology DS918+ with Portainer (1.23.2) and Docker 18.09.6 (API: 1.39). I run deCONZ as stack 

```
version: '2'
services:
  app:
    image: marthoc/deconz
    network_mode: host
    restart: always
    ports:
      - 39980:39980
      - 39981:39981
      - 39982:39982
    volumes:
      - /etc/localtime:/etc/localtime:ro
      - /volume1/docker/deCONZ/config/:/root/.local/share/dresden-elektronik/deCONZ
      #FOR DEV MODE
      - /volume1/docker/deCONZ/lib/libde_rest_plugin_motion.so:/usr/share/deCONZ/plugins/libde_rest_plugin.so
    devices:
      - /dev/ttyACM0
    environment:
      - DECONZ_WEB_PORT=18080
      - DECONZ_WS_PORT=18081
      - DEBUG_INFO=1
      - DEBUG_APS=0
      - DEBUG_ZCL=0
      - DEBUG_ZDP=0
      - DEBUG_OTAU=0
      - DECONZ_VNC_MODE=1
      - DECONZ_VNC_PORT=18082
      - DECONZ_VNC_PASSWORD=HIDDEN
      - DECONZ_UPNP=1
      - DECONZ_DEVICE=/dev/ttyACM0

```

I've compiled a new lib rest and try to mount it but it's not working ... Same logs except the lib is not loaded by the binary

**Comment by lukicsl** [comment]
@smesguich how di you compile the library? did you select the right architecture (amd64) when setting the deb sources?

**Comment by smesguich** [comment]
> 
> 
> @smesguich how di you compile the library? did you select the right architecture (amd64) when setting the deb sources?

I've compiled the lib with WSL Ubuntu 18.04.

**Comment by SAOPP** [comment]
> Hello !
> I'm facing the same issue.
> I'm running latest deCONZ docker on my Synology DS918+ with Portainer (1.23.2) and Docker 18.09.6 (API: 1.39). I run deCONZ as stack
> 
> ```
> version: '2'
> services:
>   app:
>     image: marthoc/deconz
>     network_mode: host
>     restart: always
>     ports:
>       - 39980:39980
>       - 39981:39981
>       - 39982:39982
>     volumes:
>       - /etc/localtime:/etc/localtime:ro
>       - /volume1/docker/deCONZ/config/:/root/.local/share/dresden-elektronik/deCONZ
>       #FOR DEV MODE
>       - /volume1/docker/deCONZ/lib/libde_rest_plugin_motion.so:/usr/share/deCONZ/plugins/libde_rest_plugin.so
>     devices:
>       - /dev/ttyACM0
>     environment:
>       - DECONZ_WEB_PORT=18080
>       - DECONZ_WS_PORT=18081
>       - DEBUG_INFO=1
>       - DEBUG_APS=0
>       - DEBUG_ZCL=0
>       - DEBUG_ZDP=0
>       - DEBUG_OTAU=0
>       - DECONZ_VNC_MODE=1
>       - DECONZ_VNC_PORT=18082
>       - DECONZ_VNC_PASSWORD=HIDDEN
>       - DECONZ_UPNP=1
>       - DECONZ_DEVICE=/dev/ttyACM0
> ```
> 
> I've compiled a new lib rest and try to mount it but it's not working ... Same logs except the lib is not loaded by the binary

Actually your compose looks strange, 'coz u used the host, but continue to use port mapping, for?

**Comment by smesguich** [comment]
> 
> 
> > Hello !
> > I'm facing the same issue.
> > I'm running latest deCONZ docker on my Synology DS918+ with Portainer (1.23.2) and Docker 18.09.6 (API: 1.39). I run deCONZ as stack
> > ```
> > version: '2'
> > services:
> >   app:
> >     image: marthoc/deconz
> >     network_mode: host
> >     restart: always
> >     ports:
> >       - 39980:39980
> >       - 39981:39981
> >       - 39982:39982
> >     volumes:
> >       - /etc/localtime:/etc/localtime:ro
> >       - /volume1/docker/deCONZ/config/:/root/.local/share/dresden-elektronik/deCONZ
> >       #FOR DEV MODE
> >       - /volume1/docker/deCONZ/lib/libde_rest_plugin_motion.so:/usr/share/deCONZ/plugins/libde_rest_plugin.so
> >     devices:
> >       - /dev/ttyACM0
> >     environment:
> >       - DECONZ_WEB_PORT=18080
> >       - DECONZ_WS_PORT=18081
> >       - DEBUG_INFO=1
> >       - DEBUG_APS=0
> >       - DEBUG_ZCL=0
> >       - DEBUG_ZDP=0
> >       - DEBUG_OTAU=0
> >       - DECONZ_VNC_MODE=1
> >       - DECONZ_VNC_PORT=18082
> >       - DECONZ_VNC_PASSWORD=HIDDEN
> >       - DECONZ_UPNP=1
> >       - DECONZ_DEVICE=/dev/ttyACM0
> > ```
> > 
> > 
> > I've compiled a new lib rest and try to mount it but it's not working ... Same logs except the lib is not loaded by the binary
> 
> Actually your compose looks strange, 'coz u used the host, but continue to use port mapping, for?

Well, I've comment the `network_mode:host ` (it was for a test) but the issue is the same

**Comment by lukicsl** [comment]
> -/volume1/docker/deCONZ/lib/libde_rest_plugin_motion.so:/usr/share/deCONZ/plugins/libde_rest_plugin.so

Single file volume and renaming the file, I am not sure if this works.

do a 
`docker inspect -f '{{ .Mounts }}' yor_container_id`
if the volume is mounted at all

**Comment by smesguich** [comment]
I've tried with :
/volume1/docker/deCONZ/lib/libde_rest_plugin_motion.so:/usr/share/deCONZ/plugins/libde_rest_plugin.so

```
sudo docker inspect -f '{{ .Mounts }}' deconz_app_1
[{bind  /volume1/docker/deCONZ/config /root/.local/share/dresden-elektronik/deCONZ   true rprivate} {bind  /volume1/docker/deCONZ/lib/libde_rest_plugin.so /usr/share/deCONZ/plugins/libde_rest_plugin.so   true rprivate} {bind  /etc/localtime /etc/localtime  ro false rprivate}]
```

**Comment by lukicsl** [comment]
can you you make a try 
`chmod -R a+w /volume1/docker/deCONZ`
before starting the container, just for try

**Comment by smesguich** [comment]
It's strange but it's works
(Before)
```
user@NAS:~$ ll /volume1/docker/deCONZ
total 20
drwxrwxrwx+ 1 user users   112 Apr 27 09:29 .

```

**Comment by lukicsl** [comment]
can you do:
`objdump -f libde_rest_plugin.so`
what is the output?

**Comment by smesguich** [comment]
objdump is not available inside the container. And not on my synology.

**Comment by smesguich** [comment]
Issue found :
```
Samuel@NAS:~$ sudo docker exec -it deconz_app_1 /bin/bash
root@d97f43ce8c9d:/# ls -ali /usr/share/deCONZ/plugins/
total 2572
10634 drwxr-xr-x 1 root root      126 Mar  9 14:48 .
10605 drwxr-xr-x 1 root root       84 Mar  9 14:48 ..
81255 -rwxrwxrwx 1 1029 users 2329104 Apr 28 20:02 libde_rest_plugin.so
10636 -rwxr-xr-x 1 root root    72376 Mar  8 18:19 libde_signal_plugin.so
10637 -rwxr-xr-x 1 root root   229016 Mar  8 18:19 libstd_otau_plugin.so
root@d97f43ce8c9d:/# exit
exit

Samuel@ROLON-NAS918:~$ sudo ls -ali /volume1/docker/deCONZ/lib/
total 2276
80322 drwxrwxrwx 1 Samuel users      40 Apr 28 09:21 .
76030 drwxrwxrwx 1 Samuel users     112 Apr 27 09:29 ..
81255 -rwxrwxrwx 1 Samuel users 2329104 Apr 28 20:02 libde_rest_plugin.so


```

**Comment by smesguich** [comment]
inside the conntainer with original lib : 
```
root@eb5961ae44dd:/# objdump -f /usr/share/deCONZ/plugins/libde_rest_plugin.so

/usr/share/deCONZ/plugins/libde_rest_plugin.so:     file format elf64-x86-64
architecture: i386:x86-64, flags 0x00000150:
HAS_SYMS, DYNAMIC, D_PAGED
start address 0x00000000000425d0

```

With custom lib :
```
root@494d799bff67:/# objdump -f /usr/share/deCONZ/plugins/libde_rest_plugin.so

/usr/share/deCONZ/plugins/libde_rest_plugin.so:     file format elf64-x86-64
architecture: i386:x86-64, flags 0x00000150:
HAS_SYMS, DYNAMIC, D_PAGED
start address 0x0000000000040f20

```

**Comment by smesguich** [comment]
@lukicsl Despite changing permission, It's still not working ....
IDK what could be the issue .... 

> 
> 
> Issue found :
> 
> ```
> Samuel@NAS:~$ sudo docker exec -it deconz_app_1 /bin/bash
> root@d97f43ce8c9d:/# ls -ali /usr/share/deCONZ/plugins/
> total 2572
> 10634 drwxr-xr-x 1 root root      126 Mar  9 14:48 .
> 10605 drwxr-xr-x 1 root root       84 Mar  9 14:48 ..
> 81255 -rwxrwxrwx 1 1029 users 2329104 Apr 28 20:02 libde_rest_plugin.so
> 10636 -rwxr-xr-x 1 root root    72376 Mar  8 18:19 libde_signal_plugin.so
> 10637 -rwxr-xr-x 1 root root   229016 Mar  8 18:19 libstd_otau_plugin.so
> root@d97f43ce8c9d:/# exit
> exit
> 
> Samuel@NAS:~$ sudo ls -ali /volume1/docker/deCONZ/lib/
> total 2276
> 80322 drwxrwxrwx 1 Samuel users      40 Apr 28 09:21 .
> 76030 drwxrwxrwx 1 Samuel users     112 Apr 27 09:29 ..
> 81255 -rwxrwxrwx 1 Samuel users 2329104 Apr 28 20:02 libde_rest_plugin.so
> ```

**Comment by Rick-Jongbloed** [comment]
I made the instruction from @lukicsl a oneliner :-)

`sudo apt update && sudo apt install wget git lsb-release build-essential && cd && mkdir git.com && cd git.com && git clone https://github.com/dresden-elektronik/deconz-rest-plugin.git && cd deconz-rest-plugin && wget -O - http://phoscon.de/apt/deconz.pub.key | sudo apt-key add - && sudo sh -c "echo 'deb [arch=armhf] http://phoscon.de/apt/deconz $(lsb_release -cs)-beta main' > /etc/apt/sources.list.d/deconz.list" && sudo apt update && sudo apt install deconz-dev && qmake && make -j2 && cp ../libde_rest_plugin.so /usr/share/deCONZ/plugins/ && echo "Container should reboot now...."`

Btw, I think this ticket can be closed as the original issue has been resolved. We could possible add all the information to an article on the wiki? (Like on a subsection of https://github.com/dresden-elektronik/deconz-rest-plugin/wiki/Update-deCONZ-manually)?

If the original poster still reads this ticket... @ericvandamme what do you think? Could you close this ticket?

**Comment by smesguich** [comment]
> 
> 
> I made a oneliner to compile a new version and to copy it:
> 
> `sudo apt update && sudo apt install wget git lsb-release build-essential && cd && mkdir git.com && cd git.com && git clone https://github.com/dresden-elektronik/deconz-rest-plugin.git && cd deconz-rest-plugin && wget -O - http://phoscon.de/apt/deconz.pub.key | sudo apt-key add - && sudo sh -c "echo 'deb [arch=armhf] http://phoscon.de/apt/deconz $(lsb_release -cs)-beta main' > /etc/apt/sources.list.d/deconz.list" && sudo apt update && sudo apt install deconz-dev && qmake && make -j2 && cp ../libde_rest_plugin.so /usr/share/deCONZ/plugins/ && echo "Container should reboot now...."`
> 
> Btw, I think this ticket can be closed as the original issue has been resolved. We could possible add all the information to an article on the wiki? (Like on a subsection of https://github.com/dresden-elektronik/deconz-rest-plugin/wiki/Update-deCONZ-manually)?
> 
> If the original poster still reads this ticket... @ericvandamme what do you think? Could you close this ticket?

That a not a good way to make a container, it would be better to bind a external files or include this step in the docker file in the building step image. A docker container should not compile in the lib.

**Comment by Rick-Jongbloed** [comment]
I have mounted the plugin location to a NFS share in kubernetes. This way I can compile the plugin and then destroy the container.

**Comment by smesguich** [comment]
Sorry but I still have the same issue.
I've tried an other strategy : I've tried a custom docker file based on original image.

Dockerfile
```
FROM marthoc/deconz
RUN rm -f /usr/share/deCONZ/plugins/libde_rest_plugin.so
COPY libde_rest_plugin.so /usr/share/deCONZ/plugins
RUN chown root:root /usr/bin/deCONZ
```
with
```
USER@SYNOLOGY:/volume1/docker/deCONZ/devCONZ$ ls
Dockerfile  libde_rest_plugin.so
```


```
USER@SYNOLOGY:/volume1/docker/deCONZ$ sudo docker build --tag devconz:0.0.1 devCONZ/
Password:
Sending build context to Docker daemon  2.362MB
Step 1/4 : FROM marthoc/deconz
 ---> 0d6e2f8979bc
Step 2/4 : RUN rm -f /usr/share/deCONZ/plugins/libde_rest_plugin.so
 ---> Running in b1b34765e2d1
Removing intermediate container b1b34765e2d1
 ---> 08dbc8b07402
Step 3/4 : COPY libde_rest_plugin.so /usr/share/deCONZ/plugins
 ---> aa0e54aed7e1
Step 4/4 : RUN chown root:root /usr/bin/deCONZ
 ---> Running in b1e81caf9c16
Removing intermediate container b1e81caf9c16
 ---> 56b592bb3b1a
Successfully built 56b592bb3b1a
Successfully tagged devconz:0.0.1

```

But IDK why the lib is not detected in the container logs
```
USER@SYNOLOGY:/volume1/docker/deCONZ$ sudo docker exec -it devconz_app_1 /bin/bash
Password:
root@4130559328d1:/# cd /usr/share/deCONZ/plugins/
root@4130559328d1:/usr/share/deCONZ/plugins# ls -ali
total 2612
10651 drwxr-xr-x 1 root root     126 Jun 14 13:33 .
10622 drwxr-xr-x 1 root root      84 May 26 12:39 ..
11322 -rwxr-xr-x 1 root root 2359656 Jun 14 13:30 libde_rest_plugin.so
10653 -rwxr-xr-x 1 root root   72376 May 24 20:35 libde_signal_plugin.so
10654 -rwxr-xr-x 1 root root  233496 May 24 20:35 libstd_otau_plugin.so

```

**Comment by smesguich** [comment]
Hello there,

I'll be happy to contribute but IDK why, the problems is still present ...
@marthoc any idea to make it works ?

**Comment by Mimiix** [comment]
As there hasn't been any new replies and as there are newer versions, i'm closing this issue. If this needs to be re-opened, please ping me :)
